### ISSUE TITLE
G38.2 to G38.5 positions are always in absolute machine coordinates

### ISSUE THREAD

**Issue by jannau** [Duet3D/RepRapFirmware#333]
The G38.x commands added in #324 interpret target position always absolute machine coordinates. Running 'RepRapFirmware for Duet 3 MB6HC v0.6 or 1.0 3.0beta12' on a cartesian CNC machine XYZ zero in upper left back corner. When trying following probe command sequence without triggering the probe:
```
M114
X:-10.000 Y:-10.000 Z:0.000 E:0.000 E0:0.0 E1:0.0 E2:0.0 Count -4000 -4000 0 Machine -10.000 -10.000 0.000 Bed comp 0.000
G91
G38.3 X-5 P1
M114
X:-5.000 Y:-10.000 Z:0.000 E:0.000 E0:0.0 E1:0.0 E2:0.0 Count -2000 -4000 0 Machine -5.000 -10.000 0.000 Bed comp 0.000
```
I would have expected the probe move to end at `X:-15`. When setting a workspace offset for WCS G54:
```
M114
X:0.000 Y:0.000 Z:0.000 E:0.000 E0:0.0 E1:0.0 E2:0.0 Count -4000 -4000 0 Machine -10.000 -10.000 0.000 Bed comp 0.000
G91
G38.3 X-5 P1
M114
X:5.000 Y:0.000 Z:0.000 E:0.000 E0:0.0 E1:0.0 E2:0.0 Count -2000 -4000 0 Machine -5.000 -10.000 0.000 Bed comp 0.000
```
This degrades the position arguments to a direction argument if G38.x is used in a macro since the axis' minimum or maximum are the only positions which guarantee a move in the right direction. This is also quite unsafe since it could result in a crash when the probe erroneously fails to trigger.

Is there a way to prevent further execution of a macro in the case a G38.2 or G38.4 fails? consider following macro:
```
G91
G38.2 Z-10 P1
G10 L20 P1 Z0
G90
G53 G1 Z0
```
It would be onvenient to update the Z work coordinate only if the probe triggered during the G38.2.

**Comment by T3P3** [comment]
There have been a lot of changes in RRF since this issue was raised, Please open a discussion on the forum if this is still a requirement.

**Comment by Aquilux** [comment]
I've run into this issue trying to make macros myself (3.5.0 beta 4) and had the probe move go in the wrong direction. Should I make a forum discussion?

**Comment by Aquilux** [comment]
To clarify, here is my current "probe for X to the right of current position" macro. This macro is built this way because it drastically minimizes the steps required compared to probing towards 0.

> M291 P"Ready for probe move to right?" S3
> G91                              ;switch to relative coordinates
> ;G38.2 X10                       ;probe right 10mm
> G38.2 X9999999                   ;probe right stupidly forever until this command respects G91
> G10 L20 X{-1*global.tool_Radius} ;set workplace coordinate to current position minus global.tool_Radius, use multiplication to set sign
> G0 X-10                          ;move left 10mm
> G90                              ;switch to absolute coordinates

The steps go from:
1. ensure head clears work in Z
2. position head in X
3. set WCS X=0
4. move head in X to clear work to left
5. lower Z for probe move contact
6. run macro

To:
1. move head in X to clear work to left
2. lower Z for probe move contact
3. run macro

With the G38 command ignoring the G91 command, it makes it impossible to reliably use the probe move with error checking without also running the risk of moving the wrong direction. Either you have to specify a position that will always be towards the right of the tool head no matter where it's moved and potentially not catch an error such as missing the part being probed, or you have to run the risk of missing the fact that the machine won't move as expected when you start the macro. When that's a printer head that's fine, when it's a 2lb head with a sharp cutting tool it's not as fine.

**Comment by WheelerCC** [comment]
I have a feeling I'm running into this same problem. I wonder..

https://github.com/Duet3D/RepRapFirmware/blob/87597430a5797fc8f5aa0b7672b8a07a09193a22/src/GCodes/GCodes6.cpp#L629

Should this line not account for whether or not we're in relative or absolute mode? Crudely, something like

```
float axisTarget = gb.GetDistance();
if (gb.LatestMachineState().axesRelative)
{
    axisTarget += ms.currentUserPosition[axis0]
}
else
{
    axisTarget += gb.LatestMachineState().g53Active ? ms.GetCurrentToolOffset(axis) : GetWorkplaceOffset(gb, axis); 
}
```

or something similar to https://github.com/Duet3D/RepRapFirmware/blob/87597430a5797fc8f5aa0b7672b8a07a09193a22/src/GCodes/GCodes.cpp#L2675-L2695

**Comment by dc42** [comment]
Re-opened and assign to milestone 3.6.1

**Comment by blaggacao** [comment]
Thanks for re-opening.

Is there any reliable work-around that I can use in the meantime for my custom CNC Probing Plugin?

I'm already past the realization that I can't use M585 since it also messes with the tool offset in state.

**Comment by WheelerCC** [comment]
My approach has been to use the following macro, defined once per face (front/left/back/right) to probe

```
; 0:/macros/XYZ-Probe/locate-front-edge.g

var probeOffset = 0.891         

G91                         ; relative mode
G38.2 K1 Y{move.axes[1].userPosition+5} F200
echo result
if result != 0
    abort "Didn't find an edge (fast)"
;G38.2 K1 X10 F100  ; probe negative X direction
G0 Y-1      ; retract a bit
G38.2 K1 Y{move.axes[1].userPosition+1.5} F50
if result != 0
    abort "Didn't find an edge (slow)"
G0 Y-1      ; retract a bit
G90                         ; back to absolute mode

G53 G0 Z{global.Z_MAX}
G91                         ; relative mode
G0 Y{1+var.probeOffset}      ; move to where probe thinks 0 is on relevant axis
G90                         ; back to absolute mode

```

Note I don't have this macro calculating/storing the WCS offset required to make the current tool position 0 in the given axis, I prefer to do that afterwards. This just handles the probing, and leave the tool at where the probe thinks 0 is. None the less, perhaps it can answer your question. (Note, this doesn't respect end stops so be careful)

TLDR: `move.axes[__].userPosition+_____`

**Comment by dc42** [comment]
Fixed in 3.6-dev ready for 3.6.1 release.

**Comment by WheelerCC** [comment]
I can confirm that 3.6.1 behaves as I'd expect it when using relative coordinates. 

Note that after the fix, the workaround in my macro example above needs updating (especially the right and rear equivalent, as the `move.axes[1].userPosition+5` direction and max distance will be wrong! For left and front, the direction is correct by coincidence but the max distance is still wrong) . 3.6.1 and onwards, the macro should be along the lines of

```
; 0:/macros/XYZ-Probe/locate-front-edge.g

var probeOffset = 0.891         

G91                         ; relative mode
G38.2 K1 Y5 F200
echo result
if result != 0
    abort "Didn't find an edge (fast)"
;G38.2 K1 X10 F100  ; probe negative X direction
G0 Y-1      ; retract a bit
G38.2 K1 Y1.5 F50
if result != 0
    abort "Didn't find an edge (slow)"
G0 Y-1      ; retract a bit
G90                         ; back to absolute mode

G53 G0 Z{global.Z_MAX}
G91                         ; relative mode
G0 Y{1+var.probeOffset}      ; move to where probe thinks 0 is on relevant axis
G90                         ; back to absolute mode
```
